File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

# N, M 입력
N, M = map(int, sys.stdin.readline().strip().split())

trees = list(map(int, sys.stdin.readline().strip().split()))

# 1 ~ 나무 중 가장 큰 값
start, end = 1, max(trees)

# start와 end가 같아질 때까지
# 즉, 적절한 절단기의 최대값을 찾을 떄까지 반복
while start <= end:
  
  # 중간 위치
  mid = (start + end) // 2

  # 나무 자르기
  temp = 0
  for tree in trees:
    # 나무의 높이가 절단기 높이보다 큰 경우에만
    if tree >= mid:
      # 잘라서 잘린만큼 저장
      temp += tree - mid

  # 자른 나무들의 합(temp)이 M 이상이라면
  if temp >= M:
    # 시작점 변경
    start = mid + 1
  # M 미만이라면
  else:
    # 끝점 변경
    end = mid -1

# 정답 출력
print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

// https://www.acmicpc.net/problem/2805
// 시간 복잡도: O(N * log(max(trees)))
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] trees = new int[N];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++)  trees[i] = Integer.parseInt(st.nextToken());
        Arrays.sort(trees);

        long start = 1;
        long end = trees[N-1];

        while(start <= end) {
            long mid = (start + end) / 2;
            long cnt = 0;

            for(int tree : trees){
                if(tree >= mid)     cnt += tree - mid;
            }

            if (cnt >= M)  start = mid + 1;
            else           end = mid - 1;
        }
        System.out.println(end);
    }
}